include/unifex/connect_awaitable.hpp

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/async_trace.hpp>
+#include <unifex/await_transform.hpp>
+#include <unifex/coroutine_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/coroutine.hpp>
+#include <unifex/type_traits.hpp>
+
+#if UNIFEX_NO_COROUTINES
+# error "Coroutine support is required to use <unifex/connect_awaitable.hpp>"
+#endif
+
+#include <cassert>
+#include <optional>
+#include <type_traits>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _await {
+
+template<typename Receiver>
+struct _sender_task {
+  class type;
+};
+template<typename Receiver>
+using sender_task = typename _sender_task<Receiver>::type;
+
+template<typename Receiver>
+class _sender_task<Receiver>::type {
+public:
+  struct promise_type {
+    template <typename Awaitable>
+    explicit promise_type(Awaitable&, Receiver& r) noexcept
+        : receiver_(r)
+    {}
+
+    type get_return_object() noexcept {
+      return type{
+          coro::coroutine_handle<promise_type>::from_promise(
+              *this)};
+    }
+    coro::suspend_always initial_suspend() noexcept {
+      return {};
+    }
+    [[noreturn]] coro::suspend_always final_suspend() noexcept {
+      std::terminate();
+    }
+    [[noreturn]] void unhandled_exception() noexcept {
+      std::terminate();
+    }
+    [[noreturn]] void return_void() noexcept {
+      std::terminate();
+    }
+
+    template <typename Func>
+    auto yield_value(Func&& func) noexcept {
+      struct awaiter {
+        Func&& func_;
+        bool await_ready() noexcept {
+          return false;
+        }
+        void await_suspend(coro::coroutine_handle<promise_type>) {
+          ((Func &&) func_)();
+        }
+        [[noreturn]] void await_resume() noexcept {
+          std::terminate();
+        }
+      };
+      return awaiter{(Func &&) func};
+    }
+
+    template(typename Value)
+      (requires callable<decltype(unifex::await_transform), promise_type&, Value>)
+    auto await_transform(Value&& value)
+        noexcept(is_nothrow_callable_v<decltype(unifex::await_transform), promise_type&, Value>)
+        -> callable_result_t<decltype(unifex::await_transform), promise_type&, Value> {
```

Comment:
My grasp of perfect forwarding is a bit fuzzy, but don't you need a `&&` in your `callable_result_t` here to match possibility of reference collapsing on the next line?
```suggestion
        -> callable_result_t<decltype(unifex::await_transform), promise_type&, Value&&> {
```

Reply:
It's not needed. Either `Value` is an lvalue reference, in which case the `&&` collapses away, or else it `Value` is a non-reference type, in which case calling a function with a prvalue (e.g., an `int`) is the same (semantically) as calling it with an xvalue (e.g., and `int&&`). There is no observable difference.

Reply:
I mean, I can dream up cases where it matters. E.g., if `Value&&` is `int(&&)[4]`, then calling a function with `int(&&)[4]` is different than calling it with an `int[4]`. In the latter case, the array type necessarily decays to an `int*`. But we hedge against this in the definition of `callable_result_t`, which defensively slaps a `&&` on the arguments.